src/main/java/build/buildfarm/worker/PipelineStage.java

```diff
@@ -47,24 +48,35 @@ private void runInterruptible() throws InterruptedException {
 
   @Override
   public void run() {
-    try {
-      runInterruptible();
-    } catch (InterruptedException e) {
-      // ignore
-    } catch (Exception e) {
-      getLogger()
-          .log(
-              Level.SEVERE, String.format("%s::run(): stage terminated due to exception", name), e);
-    } finally {
-      boolean wasInterrupted = Thread.interrupted();
+    boolean runStage = true;
+    while (runStage) {
       try {
-        close();
-      } finally {
-        if (wasInterrupted) {
-          Thread.currentThread().interrupt();
-        }
+        runInterruptible();
+        runStage = false;
+
+      } catch (Exception e) {
+        runStage = handleTermination(e);
       }
     }
+
+    close();
+  }
+
+  private boolean handleTermination(Exception e) {
```

Comment:
Document what the return value means here.

Reply:
improved